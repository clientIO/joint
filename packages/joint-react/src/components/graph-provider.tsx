import { GraphContext } from '../context/graph-context'
import type { dia } from '@joint/core'
import { useCreateGraphStore } from '../hooks/use-create-graph-store'

export interface GraphProps {
  /**
   * Graph instance to use. If not provided, a new graph instance will be created.
   * @see https://docs.jointjs.com/api/dia/Graph
   * @default new dia.Graph({}, { cellNamespace: shapes })
   */
  readonly graph?: dia.Graph
  /**
   * Children to render.
   */
  readonly children: React.ReactNode
  /**
   * Namespace for cell models.
   * It's loaded just once, so it cannot be used as React state.
   * @default shapes
   * @see https://docs.jointjs.com/api/shapes
   */
  readonly cellNamespace?: unknown
  /**
   * Custom cell model to use.
   * It's loaded just once, so it cannot be used as React state.
   * @see https://docs.jointjs.com/api/dia/Cell
   */
  readonly cellModel?: typeof dia.Cell
  /**
   * Initial elements to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly defaultElements?: Array<dia.Element>

  /**
   * Initial links to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly defaultLinks?: Array<dia.Link>
}

/**
 * GraphProvider component creates a graph instance and provides it to its children via context.
 * It also handles updates to the graph when cells change via React state or JointJS events.
 */
export function GraphProvider(props: GraphProps) {
  const { children, ...rest } = props
  const graphStore = useCreateGraphStore(rest)

  return <GraphContext.Provider value={graphStore}>{children}</GraphContext.Provider>
}
