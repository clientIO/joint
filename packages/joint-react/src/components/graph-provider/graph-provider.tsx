import type { dia } from '@joint/core';
import type { GraphLink } from '../../types/link-types';
import {
  GraphAreElementsMeasuredContext,
  GraphStoreContext,
} from '../../context/graph-store-context';
import { useEffect, useState, type PropsWithChildren } from 'react';
import { createStore, type Store } from '../../data/create-store';
import { useElements } from '../../hooks/use-elements';
import { useGraph } from '../../hooks';
import { setLinks } from '../../utils/cell/set-cells';
import type { GraphElement } from '../../types/element-types';

interface GraphProviderHandlerProps {
  /**
   * Initial links to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly initialLinks?: Array<dia.Link | GraphLink>;
}

/**
 * GraphProviderHandler component is used to handle the graph instance and provide it to the children.
 * It also handles the default elements and links.
 * @param props - {GraphProviderHandler} props
 * @param props.children - Children to render.
 * @param props.initialLinks - Initial links to be added to graph
 * @returns GraphProviderHandler component
 * @private
 */
function GraphProviderHandler({
  children,
  initialLinks,
}: PropsWithChildren<GraphProviderHandlerProps>) {
  const areElementsMeasured = useElements((items) => {
    let areMeasured = true;
    for (const [, { width = 0, height = 0 }] of items) {
      if (width <= 1 || height <= 1) {
        areMeasured = false;
        break;
      }
    }
    return areMeasured;
  });
  const graph = useGraph();

  useEffect(() => {
    if (areElementsMeasured) {
      setLinks({ graph, initialLinks });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [areElementsMeasured, graph]);

  return (
    <GraphAreElementsMeasuredContext.Provider value={areElementsMeasured}>
      {children}
    </GraphAreElementsMeasuredContext.Provider>
  );
}

export interface GraphProps {
  /**
   * Graph instance to use. If not provided, a new graph instance will be created.
   * @see https://docs.jointjs.com/api/dia/Graph
   * @default new dia.Graph({}, { cellNamespace: shapes })
   */
  readonly graph?: dia.Graph;
  /**
   * Children to render.
   */
  readonly children?: React.ReactNode;
  /**
   * Namespace for cell models.
   * It's loaded just once, so it cannot be used as React state.
   * When added new shape, it will not remove existing ones, it will just add new ones.
   * So `{ ...shapes, ReactElement }` elements are still available.
   * @default  `{ ...shapes, ReactElement }`
   * @see https://docs.jointjs.com/api/shapes
   */
  readonly cellNamespace?: unknown;
  /**
   * Custom cell model to use.
   * It's loaded just once, so it cannot be used as React state.
   * @see https://docs.jointjs.com/api/dia/Cell
   */
  readonly cellModel?: typeof dia.Cell;
  /**
   * Initial elements to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly initialElements?: Array<dia.Element | GraphElement>;
  /**
   * Initial links to be added to graph
   * It's loaded just once, so it cannot be used as React state.
   */
  readonly initialLinks?: Array<dia.Link | GraphLink>;

  /**
   * Store is build around graph, it handles react updates and states, it can be created separately and passed to the provider via `createStore` function.
   * @see `createStore`
   */
  readonly store?: Store;
}

/**
 *
 * GraphProvider component creates a graph instance and provide `dia.graph` to it's children.
 * It relies on @see useCreateGraphStore hook to create the graph instance.
 *
 * Without this provider, the library will not work.
 * @param props - {GraphProvider} props
 * @returns GraphProvider component
 * @example
 * Using provider:
 * ```tsx
 * import { GraphProvider } from '@joint/react'
 *
 * function App() {
 *  return (
 *   <GraphProvider>
 *    <MyApp />
 *  </GraphProvider>
 * )
 * ```
 * @example
 * Using provider with default elements and links:
 * ```tsx
 * import { GraphProvider } from '@joint/react'
 *
 * function App() {
 *  return (
 *   <GraphProvider initialElements={[]} initialLinks={[]}>
 *    <MyApp />
 *  </GraphProvider>
 * )
 * ```
 * @group Components
 */
export function GraphProvider(props: Readonly<GraphProps>) {
  const { children, initialLinks, store, ...rest } = props;

  /**
   * Graph store instance.
   * @returns - The graph store instance.
   */

  const [graphStore, setGraphStore] = useState<null | Store>(null);

  useEffect(() => {
    const newStore = store ?? createStore({ ...rest });
    // We must use state initialization for the store, because it can be used in the same component.
    // eslint-disable-next-line @eslint-react/hooks-extra/no-direct-set-state-in-use-effect
    setGraphStore(newStore);
    return () => {
      if (newStore) {
        newStore.destroy();
      }
    };
    // On load initialization
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  if (graphStore === null) {
    return null;
  }

  return (
    <GraphStoreContext.Provider value={graphStore}>
      <GraphProviderHandler initialLinks={initialLinks}>{children}</GraphProviderHandler>
    </GraphStoreContext.Provider>
  );
}
