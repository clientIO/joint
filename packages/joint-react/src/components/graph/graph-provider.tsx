import type { dia } from '@joint/core';
import type { GraphLink } from '../../types/link-types';
import {
  forwardRef,
  useLayoutEffect,
  type Dispatch,
  type PropsWithChildren,
  type SetStateAction,
} from 'react';
import { createStore, type GraphStore } from '../../data/create-graph-store';
import { useElements } from '../../hooks/use-elements';
import { useGraphStore } from '../../hooks';
import type { GraphElement } from '../../types/element-types';
import { CONTROLLED_MODE_BATCH_NAME } from '../../utils/graph/update-graph';
import { useImperativeApi } from '../../hooks/use-imperative-api';
import { GraphAreElementsMeasuredContext, GraphStoreContext } from '../../context';
import { getTargetOrSource } from '../../utils/cell/get-link-targe-and-source-ids';

interface GraphProviderBaseProps<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Element extends dia.Element | GraphElement = any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Link extends dia.Link | GraphLink = any,
> {
  /**
   * Elements (nodes) to be added to graph.
   * When `onElementsChange`, it enabled controlled mode.
   * If there is no `onElementsChange` provided, it will be used just on onload (initial)
   */
  readonly elements?: Element[];

  /**
   * Links (edges) to be added to graph.
   * When `onLinksChange`, it enabled controlled mode.
   * If there is no `onLinksChange` provided, it will be used just on onload (initial)
   */
  readonly links?: Link[];

  /**
   * Callback triggered when elemeÂ§nts (nodes) change.
   * Providing this prop enables controlled mode for elements.
   * If specified, this function will override the default behavior, allowing you to manage all element changes manually instead of relying on `graph.change`.
   */
  readonly onElementsChange?: Dispatch<SetStateAction<Element[]>>;

  /**
   * Callback triggered when links (edges) change.
   * Providing this prop enables controlled mode for links.
   * If specified, this function will override the default behavior, allowing you to manage all link changes manually instead of relying on `graph.change`.
   */
  readonly onLinksChange?: Dispatch<SetStateAction<Link[]>>;
}

/**
 * Internal handler coordinating initial population and controlled-mode mirroring
 * for elements and links. Also delays link creation until elements are measured
 * in uncontrolled mode to avoid flicker.
 * @param props - The properties for the GraphProviderHandler, including elements, links, and callbacks.
 * @returns A context provider for the measured state of elements.
 * @private
 */
export function GraphProviderHandler<
  Element extends dia.Element | GraphElement = dia.Element,
  Link extends dia.Link | GraphLink = dia.Link,
>(props: PropsWithChildren<GraphProviderBaseProps<Element, Link>>) {
  const { elements, links, onElementsChange, onLinksChange, children } = props;
  const areElementsMeasured = useElements((items) => {
    let areMeasured = true;
    for (const { width = 0, height = 0 } of items) {
      if (width <= 1 || height <= 1) {
        areMeasured = false;
        break;
      }
    }
    return areMeasured;
  });

  const { graph, setElements, setLinks } = useGraphStore();

  const areElementsInControlledMode = !!onElementsChange;
  const areLinksInControlledMode = !!onLinksChange;
  const isControlledMode = areElementsInControlledMode || areLinksInControlledMode;
  // Controlled mode for elements
  useLayoutEffect(() => {
    if (!areElementsMeasured) return;
    if (!graph) return;
    if (!isControlledMode) return;

    graph.startBatch(CONTROLLED_MODE_BATCH_NAME);
    if (areElementsInControlledMode && elements !== undefined) {
      setElements(elements);
    }
    if (areLinksInControlledMode && links !== undefined) {
      setLinks(links as GraphLink[]);
    }
    graph.stopBatch(CONTROLLED_MODE_BATCH_NAME);
  }, [
    areElementsInControlledMode,
    areElementsMeasured,
    areLinksInControlledMode,
    graph,
    elements,
    links,
    isControlledMode,
    setElements,
    setLinks,
  ]);

  useLayoutEffect(() => {
    // with this all links are connected only when react elements are measured
    // It fixes issue with a flickering of un-measured react elements.
    if (isControlledMode) return;
    if (!areElementsMeasured) return;
    if (!links?.length) return;
    const hasSomePort = links?.some((link) => {
      const { source, target } = link;
      const sourceObject = getTargetOrSource(source as dia.Link.EndJSON);
      const targetObject = getTargetOrSource(target as dia.Link.EndJSON);
      return sourceObject.port || targetObject.port;
    });
    if (!hasSomePort) return;
    setLinks(links as GraphLink[]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [areElementsMeasured, isControlledMode]);

  return (
    <GraphAreElementsMeasuredContext.Provider value={areElementsMeasured}>
      {children}
    </GraphAreElementsMeasuredContext.Provider>
  );
}

export interface GraphProps<
  Graph extends dia.Graph = dia.Graph,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Element extends dia.Element | GraphElement = any,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Link extends dia.Link | GraphLink = any,
> extends GraphProviderBaseProps<Element, Link> {
  /**
   * Graph instance to use. If not provided, a new graph instance will be created.
   * @see https://docs.jointjs.com/api/dia/Graph
   * @default new dia.Graph({}, { cellNamespace: shapes })
   */
  readonly graph?: Graph;
  /**
   * Children to render.
   */
  readonly children?: React.ReactNode;
  /**
   * Namespace for cell models.
   * It's loaded just once, so it cannot be used as React state.
   * When added new shape, it will not remove existing ones, it will just add new ones.
   * So `{ ...shapes, ReactElement }` elements are still available.
   * @default  `{ ...shapes, ReactElement }`
   * @see https://docs.jointjs.com/api/shapes
   */
  readonly cellNamespace?: unknown;
  /**
   * Custom cell model to use.
   * It's loaded just once, so it cannot be used as React state.
   * @see https://docs.jointjs.com/api/dia/Cell
   */
  readonly cellModel?: typeof dia.Cell;

  /**
   * Store is build around graph, it handles react updates and states, it can be created separately and passed to the provider via `createStore` function.
   * @see `createStore`
   */
  readonly store?: GraphStore<Graph>;
}

/**
 * Graph component creates a graph instance and provides `dia.graph` to its children.
 * This component is essential for the library to function correctly. It manages the graph instance and supports controlled and uncontrolled modes for elements and links.
 * @param props - The properties for the Graph component.
 * @param forwardedRef - A reference to the GraphStore instance.
 * @returns The Graph component.
 * @example
 * Using the Graph component:
 * ```tsx
 * import { Graph } from '@joint/react';
 * function App() {
 *   return (
 *     <Graph>
 *       <MyApp />
 *     </Graph>
 *   );
 * }
 * ```
 * @example
 * Using the Graph component with default elements and links:
 * ```tsx
 * import { Graph } from '@joint/react';
 * function App() {
 *   return (
 *     <Graph elements={[]} links={[]}>
 *       <MyApp />
 *     </Graph>
 *   );
 * }
 * ```
 */
function GraphBase<Element extends dia.Element | GraphElement, Link extends dia.Link | GraphLink>(
  props: Readonly<GraphProps<dia.Graph, Element, Link>>,
  forwardedRef: React.Ref<GraphStore>
) {
  const { children, store, ...rest } = props;
  /**
   * Graph store instance.
   * @returns - The graph store instance.
   */

  const { isReady, ref } = useImperativeApi<GraphStore>(
    {
      forwardedRef,
      onLoad() {
        const newStore = store ?? createStore({ ...rest });
        // We must use state initialization for the store, because it can be used in the same component.

        return {
          cleanup() {
            if (newStore) {
              newStore.destroy(!!rest.graph || !!store?.graph);
            }
          },
          instance: newStore,
        };
      },
    },
    []
  );

  if (!isReady) {
    return null;
  }

  return (
    <GraphStoreContext.Provider value={ref.current}>
      <GraphProviderHandler {...props}>{children}</GraphProviderHandler>
    </GraphStoreContext.Provider>
  );
}

/**
 * GraphProviderHandler component is used to handle the graph instance and provide it to the children.
 * It also handles the default elements and links.
 * @returns GraphProviderHandler component
 * @param props - {GraphProviderHandler} props
 * @private
 */
export const GraphProvider = forwardRef(GraphBase) as <
  Element extends dia.Element | GraphElement = dia.Element,
  Link extends dia.Link | GraphLink = dia.Link,
>(
  props: Readonly<GraphProps<dia.Graph, Element, Link>> & {
    ref?: React.Ref<GraphStore>;
  }
) => ReturnType<typeof GraphBase>;
