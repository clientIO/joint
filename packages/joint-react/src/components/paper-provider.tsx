import type { ReactNode } from 'react'
import { useState } from 'react'
import type { PaperOptions } from '../utils/create-paper'
import { createPaper } from '../utils/create-paper'
import { PaperContext } from '../context/paper-context'
import type { dia } from '@joint/core'
import { useGraph } from '../hooks/use-graph'

export interface PaperProviderProps extends PaperOptions {
  readonly children: ReactNode
}

/**
 * Paper provider creates a paper instance and provides it to its children.
 * It extends the paper options from the createPaper function.
 * For more information about paper, see the JointJS documentation.
 * @see https://docs.jointjs.com/api/dia/Paper
 */
export function PaperProvider({ children, ...paperOptions }: Readonly<PaperProviderProps>) {
  const graph = useGraph()
  if (!graph) {
    throw new Error('PaperProvider must be used within a GraphProvider')
  }

  const [paper] = useState<dia.Paper>(() => createPaper(graph, paperOptions))
  return <PaperContext.Provider value={paper}>{children}</PaperContext.Provider>
}
