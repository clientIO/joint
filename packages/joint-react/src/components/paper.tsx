import { mvc, type dia } from '@joint/core'
import { useCallback, useEffect, useRef, useState, type CSSProperties, type ReactNode } from 'react'
import { usePaper } from '../hooks/use-paper'
import { useGraph } from '../hooks/use-graph'

export const PORTAL_READY_EVENT = 'portal:ready'

/**
 * The props for the Paper component. Extend the `dia.Paper.Options` interface.
 * For more information, see the JointJS documentation.
 * @see https://docs.jointjs.com/api/dia/Paper
 */
export interface PaperProps extends dia.Paper.Options {
  /**
   * A function that renders the element. It is called every time the element is rendered.
   */
  renderElement?: (_element: dia.Element) => ReactNode
  /**
   * A function that is called when the paper is ready.
   * @param paper The JointJS paper instance.
   */
  onReady?: (paper: dia.Paper) => void
  /**
   * A function that is called when an event is triggered on the paper.
   * @param paper The JointJS paper instance.
   * @param args The arguments passed to the event.
   */
  onEvent?: (paper: dia.Paper, ...args: unknown[]) => void
  /**
   * The style of the paper element.
   */
  style?: CSSProperties
  /**
   * Class name of the paper element.
   */
  className?: string
  /**
   * The data attributes to listen to changes.
   */
  dataAttributes?: string[]
  /**
   * The selector of the portal element.
   */
  portalSelector?: string | ((view: dia.ElementView) => HTMLElement | null)
}

type ElementState = {
  id: dia.Cell.ID // ID of the element
  data: unknown // Data from the `model.get('data')`
  containerEl: HTMLElement | null // Reference to the container element
}

/**
 * Paper component that renders the JointJS paper element.
 */
export function Paper(props: Readonly<PaperProps>) {
  const {
    // renderElement,
    onReady,
    onEvent,
    style,
    className,
    dataAttributes = ['data'],
    // portalSelector = 'portal',
    ...paperOptions
  } = props

  const paperWrapperElementRef = useRef<HTMLDivElement | null>(null)
  const paper = usePaper(paperOptions)
  const graph = useGraph()

  // TODO - this is not used yet
  const [elements, setElements] = useState<Record<string, ElementState>>({})
  const elementPortals = Object.values(elements).map((element) => {
    return <div key={element.id}>{/* {renderElement ? renderElement(element) : null} */}</div>
  })
  // function resizePaperWrapper() {
  //   if (paperWrapperElementRef.current) {
  //     paperWrapperElementRef.current.style.width = paper.el.style.width
  //     paperWrapperElementRef.current.style.height = paper.el.style.height
  //   }
  // }

  const resizePaperWrapper = useCallback(() => {
    if (paperWrapperElementRef.current) {
      paperWrapperElementRef.current.style.width = paper.el.style.width
      paperWrapperElementRef.current.style.height = paper.el.style.height
    }
  }, [paper])

  function setElement(model: dia.Cell, containerElement?: HTMLElement): void {
    setElements((previousState) => {
      const { id } = model
      return {
        ...previousState,
        [id]: {
          id,
          data: model.get('data'),
          containerEl: containerElement ?? previousState[id]?.containerEl ?? null,
        },
      }
    })
  }

  // const setElementData = useCallback((id: dia.Cell.ID, data: unknown) => {
  //   setElements((previousState) => {
  //     return {
  //       ...previousState,
  //       [id]: {
  //         ...previousState[id],
  //         data,
  //       },
  //     }
  //   })
  // }, [])

  const bindEvents = useCallback(() => {
    // An object to keep track of the listeners. It's not exposed, so the users
    // can't undesirably remove the listeners.
    const controller = new mvc.Listener()

    // Update the elements state when the graph data changes
    const attributeChangeEvents = dataAttributes.map((attribute) => `change:${attribute}`).join(' ')

    controller.listenTo(graph, attributeChangeEvents, setElement)
    // Update the portal node reference when the element view is rendered
    controller.listenTo(paper, PORTAL_READY_EVENT, (elementView, portalElement) =>
      setElement(elementView.model, portalElement)
    )

    controller.listenTo(paper, 'resize', resizePaperWrapper)

    if (onEvent) {
      controller.listenTo(paper, 'all', (...args) => onEvent(paper, ...args))
    }

    return () => controller.stopListening()
  }, [dataAttributes, graph, onEvent, paper, resizePaperWrapper])

  useEffect(() => {
    paperWrapperElementRef.current?.append(paper.el)
    resizePaperWrapper()
    paper.unfreeze()

    const unbindEvents = bindEvents()
    if (onReady) {
      onReady(paper)
    }

    return () => {
      // I am not sure about removing the paper
      // paper.remove()
      paper.freeze()
      unbindEvents()
    }
  }, [bindEvents, dataAttributes, graph, onEvent, onReady, paper, resizePaperWrapper]) // options, onReady, onEvent, style
  return (
    <div className={className} ref={paperWrapperElementRef} style={style}>
      {elementPortals}
    </div>
  )
}
