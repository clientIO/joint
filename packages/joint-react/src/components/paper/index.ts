export * from './paper';
export * from './paper.types';
export * from './render-element/paper-element-item';
export * from './render-element/paper-html-container';
