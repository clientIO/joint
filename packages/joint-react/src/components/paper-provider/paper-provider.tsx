import { useContext, useEffect, useMemo, useState } from 'react';
import { GraphStoreContext, PaperContext } from '../../context';
import { dia } from '@joint/core';
import { useGraph } from '../../hooks';
import { createPortsStore } from '../../data/create-ports-store';
import type { PortElementsCacheEntry } from '../../data/create-ports-data';
import type { OmitWithoutIndexSignature } from '../../types';
import { GraphProvider, type GraphProps } from '../graph-provider/graph-provider';
import { usePaperElementRenderer } from '../../hooks/use-paper-element-renderer';
const DEFAULT_CLICK_THRESHOLD = 10;

export type OnPaperRenderElement = (element: dia.Element, portalElement: SVGElement) => void;

export type ReactPaperOptions = OmitWithoutIndexSignature<dia.Paper.Options, 'frozen'>;

// Interface for Paper options, extending JointJS Paper options
export interface PaperOptions extends ReactPaperOptions {
  readonly scale?: number;
  /**
   * A function that is called when the paper is ready.
   * @param element - The element that is being rendered
   * @param portalElement  - The portal element that is being rendered
   * @returns
   */
  readonly onRenderElement?: OnPaperRenderElement;
}

export interface PaperProviderProps extends ReactPaperOptions, GraphProps {
  readonly children: React.ReactNode;
}
const EMPTY_OBJECT = {} as const;

// eslint-disable-next-line jsdoc/require-jsdoc
function Component(props: PaperProviderProps) {
  const { children, ...paperOptions } = props;
  const graph = useGraph();

  const { onRenderElement, recordOfSVGElements } = usePaperElementRenderer();

  const [paperCtx] = useState<PaperContext>(function (): PaperContext {
    const portsStore = createPortsStore();
    const elementView = dia.ElementView.extend({
      // Render element using react, `elementView.el` is used as portal gate for react (createPortal)
      onRender() {
        // eslint-disable-next-line unicorn/no-this-assignment, @typescript-eslint/no-this-alias, no-shadow, @typescript-eslint/no-shadow
        const elementView: dia.ElementView = this;
        onRenderElement(elementView.model, elementView.el as SVGGElement);
      },
      // Render port using react, `portData.portElement.node` is used as portal gate for react (createPortal)
      _renderPorts() {
        // This is firing when the ports are rendered (updated, inserted, removed)
        // @ts-expect-error we use private jointjs api method, it throw error here.
        dia.ElementView.prototype._renderPorts.call(this);
        // eslint-disable-next-line unicorn/no-this-assignment, @typescript-eslint/no-this-alias, no-shadow, @typescript-eslint/no-shadow
        const elementView: dia.ElementView = this;

        const portElementsCache: Record<string, PortElementsCacheEntry> = this._portElementsCache;
        portsStore.onRenderPorts(elementView.model.id, portElementsCache);
      },
    });

    // Create a new JointJS Paper with the provided options
    const paper = new dia.Paper({
      async: true,
      sorting: dia.Paper.sorting.APPROX,
      preventDefaultBlankAction: false,
      frozen: true,
      model: graph,
      elementView,
      ...paperOptions,
      clickThreshold: paperOptions.clickThreshold ?? DEFAULT_CLICK_THRESHOLD,
    });

    return {
      paper,
      portsStore,
      recordOfSVGElements: EMPTY_OBJECT,
    };
  });

  useEffect(() => {
    paperCtx.paper.options = {
      ...paperCtx.paper.options,
      ...paperOptions,
    };
  }, [paperCtx.paper, paperOptions]);

  const paperContextFull = useMemo((): PaperContext => {
    return {
      ...paperCtx,
      recordOfSVGElements,
    };
  }, [paperCtx, recordOfSVGElements]);

  // Remove the check for existing context, always provide PaperContext
  return <PaperContext.Provider value={paperContextFull}>{children}</PaperContext.Provider>;
}

/**
 * PaperProvider is a React component that provides a context for managing the state of the paper.
 * It uses the PaperContext to provide a value to its children.
 * The context value is an array containing the current paper context and a function to update it.
 * @param props - The props object containing the children components.
 * @param props.children - The children components that will have access to the PaperContext.
 * @returns - A JSX element that wraps the children with the PaperContext provider.
 * @group Components
 */
export function PaperProvider(props: PaperProviderProps) {
  const {
    children,
    initialElements,
    initialLinks,
    graph,
    cellNamespace,
    cellModel,
    store,
    ...paperOptions
  } = props;
  const hasStore = !!useContext(GraphStoreContext);
  const content = <Component {...paperOptions}>{children}</Component>;
  if (hasStore) {
    return content;
  }
  return (
    <GraphProvider
      initialElements={initialElements}
      initialLinks={initialLinks}
      graph={graph}
      cellNamespace={cellNamespace}
      cellModel={cellModel}
      store={store}
    >
      {content}
    </GraphProvider>
  );
}
