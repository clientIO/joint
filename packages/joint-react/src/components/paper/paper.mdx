import { Meta, Story, Canvas, Controls, Markdown } from '@storybook/blocks';
import * as Stories from './paper.stories';

<Meta of={Stories} />

# Paper Component

Paper component that renders the JointJS elements (nodes) inside HTML.
It uses `renderElement` to render the elements.

For more about native paper component, see [paper](https://docs.jointjs.com/learn/quickstart/paper).

    <Canvas of={Stories.WithRectElement} />
    <Canvas of={Stories.WithHtmlElement} />
<Controls/>
