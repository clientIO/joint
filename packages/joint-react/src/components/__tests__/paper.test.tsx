/* eslint-disable react-perf/jsx-no-new-array-as-prop */
import { render, waitFor } from '@testing-library/react'
import { GraphProvider } from '../graph-provider'
import { PaperProvider } from '../paper-provider'
import { Paper } from '../paper'

describe('paper', () => {
  it('should render paper component without errors', async () => {
    const { asFragment, getByText } = render(
      <GraphProvider
        cells={[
          {
            id: '1',
            type: 'standard.Rectangle',
            position: { x: 10, y: 100 },
            attrs: { label: { text: 'testing-rectangle1' } },
          },
          {
            id: '2',
            type: 'standard.Rectangle',
            position: { x: 10, y: 100 },
            attrs: { label: { text: 'testing-rectangle2' } },
          },
          {
            type: 'standard.Link',
            source: { id: '1' },
            target: { id: '2' },
          },
        ]}
      >
        <PaperProvider>
          <Paper />
        </PaperProvider>
      </GraphProvider>
    )

    await waitFor(async () => {
      expect(getByText('testing-rectangle1')).toBeDefined()
      expect(getByText('testing-rectangle2')).toBeDefined()
      const fragment = asFragment()
      // remove model-id as it's auto-generated id from jointjs/core
      for (const element of fragment.querySelectorAll('[model-id]')) {
        element.removeAttribute('model-id')
      }
      expect(fragment).toMatchSnapshot()
    })
  })

  it('should call onReady with the paper instance upon mount', async () => {
    const onReadyMock = jest.fn()
    render(
      <GraphProvider>
        <PaperProvider>
          <Paper onReady={onReadyMock} />
        </PaperProvider>
      </GraphProvider>
    )
    await waitFor(() => {
      expect(onReadyMock).toHaveBeenCalledTimes(1)
      expect(onReadyMock.mock.calls[0][0].el).toBeDefined()
    })
  })
})
