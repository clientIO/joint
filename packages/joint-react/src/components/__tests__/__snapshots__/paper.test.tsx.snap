// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`paper should render paper component without errors 1`] = `
<DocumentFragment>
  <div
    style="width: 100%; height: 100%;"
  >
    <div
      class="joint-paper joint-theme-default"
      style="position: relative; width: 100%; height: 100%;"
    >
      <div
        class="joint-paper-background"
        joint-selector="background"
        style="position: absolute; inset: 0;"
      />
      <svg
        height="100%"
        id="v-2"
        joint-selector="svg"
        style="position: absolute; inset: 0; overflow: hidden;"
        width="100%"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <style
          id="v-3"
          type="text/css"
        />
        <defs
          joint-selector="defs"
        >
          <marker
            id="v-21801458145"
            markerUnits="userSpaceOnUse"
            orient="auto"
            overflow="visible"
          >
            <path
              d="M 10 -5 0 0 10 5 z"
              fill="#333333"
              id="v-11"
              stroke="#333333"
              transform="rotate(180)"
            />
          </marker>
        </defs>
        <g
          class="joint-layers"
          joint-selector="layers"
        >
          <g
            class="joint-grid-layer"
            style="pointer-events: none;"
          />
          <g
            class="joint-back-layer"
          />
          <g
            class="joint-cells-layer joint-viewport"
            style="webkit-user-select: none; user-select: none;"
          >
            <g
              class="joint-cell joint-type-standard joint-type-standard-rectangle joint-element joint-theme-default"
              cursor="move"
              data-type="standard.Rectangle"
              id="j_1"
              transform="translate(10,100)"
            >
              <rect
                fill="#FFFFFF"
                height="1"
                id="v-4"
                joint-selector="body"
                stroke="#000000"
                stroke-width="2"
                width="1"
              />
              <text
                fill="#333333"
                font-size="14"
                id="v-5"
                joint-selector="label"
                text-anchor="middle"
                x="0.5"
                xml:space="preserve"
                y="0.5"
              >
                <tspan
                  class="v-line"
                  dy="0.3em"
                >
                  testing-rectangle1
                </tspan>
              </text>
            </g>
            <!--z-index:2-->
            <g
              class="joint-cell joint-type-standard joint-type-standard-rectangle joint-element joint-theme-default"
              cursor="move"
              data-type="standard.Rectangle"
              id="j_2"
              transform="translate(10,100)"
            >
              <rect
                fill="#FFFFFF"
                height="1"
                id="v-6"
                joint-selector="body"
                stroke="#000000"
                stroke-width="2"
                width="1"
              />
              <text
                fill="#333333"
                font-size="14"
                id="v-7"
                joint-selector="label"
                text-anchor="middle"
                x="0.5"
                xml:space="preserve"
                y="0.5"
              >
                <tspan
                  class="v-line"
                  dy="0.3em"
                >
                  testing-rectangle2
                </tspan>
              </text>
            </g>
            <!--z-index:3-->
            <g
              class="joint-cell joint-type-standard joint-type-standard-link joint-link joint-theme-default"
              data-type="standard.Link"
              id="j_4"
            >
              <path
                cursor="pointer"
                d="M 0 0 L 0 0"
                fill="none"
                id="v-9"
                joint-selector="wrapper"
                stroke="transparent"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="10"
              />
              <path
                d="M 0 0 L 0 0"
                fill="none"
                id="v-8"
                joint-selector="line"
                marker-end="url(#v-21801458145)"
                pointer-events="none"
                stroke="#333333"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </g>
            <!--z-index:4-->
          </g>
          <g
            class="joint-labels-layer joint-viewport"
            style="webkit-user-select: none; user-select: none;"
          />
          <g
            class="joint-front-layer"
          />
          <g
            class="joint-tools-layer"
          />
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
